plugin/minimal/minimal.go

```diff
@@ -0,0 +1,37 @@
+package minimal
+
+import (
+	"context"
+	"time"
+
+	"github.com/coredns/coredns/plugin"
+	"github.com/coredns/coredns/plugin/pkg/nonwriter"
+	"github.com/coredns/coredns/plugin/pkg/response"
+	"github.com/miekg/dns"
+)
+
+// minimalHandler implements the plugin.Handler interface.
+type minimalHandler struct {
+	Next plugin.Handler
+}
+
+func (m *minimalHandler) Name() string { return "minimal" }
+
+func (m *minimalHandler) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
+	nw := nonwriter.New(w)
+
+	rcode, err := plugin.NextOrFailure(m.Name(), m.Next, ctx, nw, r)
+	if err != nil {
+		return rcode, err
+	}
+	w.WriteMsg(m.minimizeResponse(nw.Msg))
+	return rcode, err
+}
+
+func (m *minimalHandler) minimizeResponse(msg *dns.Msg) *dns.Msg {
+	if ty, _ := response.Typify(msg, time.Now().UTC()); ty == response.NoError {
```

Comment:
you need to create a new message here to prevent modifying an existing dns.Msg that we cache for instance, also worth changing this into a switch statement and at least have all the response.* handled there (some will be empty).

Also this function can be removed and the code moved into the `ServeDNS` function

Reply:
instead of mentioning each `response.*` type as a case would it make more sense to keep one case for `response.NoError` and a default case (empty) for others? currently the other cases don't have any logic so until new cases are required we can use the default case directly.